docs/design/endpoints.md

```diff
@@ -117,41 +296,53 @@ public open class EndpointProviderGenerator(
 }
 ```
 
 ### Middleware
-Endpoint resolution, whether generated or written by the caller, is performed per-request as a pre-operation middleware.
-At runtime, endpoint resolution is comprised of two principal tasks:
-1. construct a set of provider parameters, sourcing the values as specified by the model (context params)
-2. using the sourced parameters, call the provider to resolve to an endpoint, pointing the request to it
 
-Since endpoint resolution is performed on a per-operation basis, and the sourcing of provider parameters can vary
-per-operation, middleware must now be generated per-operation (this differs from the original implementation, where the
-endpoint resolver middleware was static across operations and its implementation was part of the runtime).
+Endpoint resolution, whether generated or written by the caller, is performed per-request as a pre-serialization
+middleware. At runtime, endpoint resolution is comprised of two principal tasks:
+1. construct a set of provider parameters, binding the values as specified by the model (context params)
+2. using the bound parameters, call the provider to resolve to an endpoint, pointing the request to it
+
+Operations that don't influence provider parameter values (i.e. operations with no associated `staticContextParams`
+or `clientContextParams`) can use a default middleware generated once for the service. Other operations will need to
+have unique middlewares generated to bind those specific values.
 
 ```kotlin
 public open class EndpointResolverMiddlewareGenerator(
     public val writer: KotlinWriter,
     public val operation: OperationShape,
-    public val ruleset: EndpointRuleset,
+    public val ruleSet: EndpointRuleset,
     // will likely need some additional params such as service shape etc.
 ) {
     /**
      * Generates code to construct a set of provider parameters based on the specific service, call the provider
      * with those parameters, and modify the request (and handle failure to resolve if that occurs).
      * 
-     * Client/static/general context parameters are inspected here to determine where to source values.
+     * Client/static/general context parameters are inspected here to determine how to bind values.
      */
     public fun render()
+
+    /**
+     * Implement this API to generate additional parameters to be passed to the middleware.
+     * By default, the middleware takes two parameters:
+     * 1. service client config
+     * 2. operation's request object
+     * 
+     * The implementer is responsible for code generation of the middleware installation.
+     */
+    public open fun getMiddlewareParams(): List<Symbol>
     
     /**
-     * SDK authors extend this method to handle builtins.
+     * SDK authors extend this method to handle built-ins.
+     * Called within the builder lambda for the parameters object.
      */
-    public open fun resolveBuiltin(value: String)
+    public open fun renderBuiltin(name: String)
```

Comment:
Same question as above around `resolveFunction`, perhaps I'm just not seeing the bigger picture here yet...